src/Shell/Task/ModelTask.php

```diff
@@ -344,25 +344,56 @@ public function findHasMany($model, array $associations)
                 }
             }
 
-            foreach ($otherSchema->columns() as $fieldName) {
-                $assoc = false;
-                if (!in_array($fieldName, $primaryKey) && $fieldName === $foreignKey) {
-                    $assoc = [
-                        'alias' => $otherModel->alias(),
-                        'foreignKey' => $fieldName
-                    ];
-                } elseif ($otherTableName === $tableName && $fieldName === 'parent_id') {
-                    $className = ($this->plugin) ? $this->plugin . '.' . $model->alias() : $model->alias();
-                    $assoc = [
-                        'alias' => 'Child' . $model->alias(),
-                        'className' => $className,
-                        'foreignKey' => $fieldName
-                    ];
+            $targetForeignKey = $primaryKey;
+            foreach ($targetForeignKey as $i => $fieldname) {
+                if ($fieldname === 'id') {
+                    $targetForeignKey[$i] = Inflector::singularize($model->table()) . '_id';
                 }
-                if ($assoc && $this->plugin && empty($assoc['className'])) {
-                    $assoc['className'] = $this->plugin . '.' . $assoc['alias'];
+            }
+
+            $possibleForeignKeys = $otherSchema->columns();
```

Comment:
Looking at that, thoroughly, suggests to me that hasMany detection can fail (see ProjectsTable which misses some).